File: src/Pybind11Wraps/Utilities/uniform_random_01.py
#-------------------------------------------------------------------------------
# uniform_random_01
#-------------------------------------------------------------------------------
from PYB11Generator import *

class uniform_random_01:
    "Encapsulate a random number generator to generate numbers in [0,1)."

    #...........................................................................
    # Constructors
    def pyinit(self):
        "Default constructor"

    def pyinit1(self, seed="const size_t"):
        "Construct with a seed value"

    #...........................................................................
    # Methods
    def __call__(self):
        "Return a random value in the range [0,1)"
        return "double"

    def seed(self, val="const size_t"):
        "Set the seed for our generator"
        return "void"

    def advance(self, n="const size_t"):
        "Advance the state of the generator as though called n times"
        return "void"

    def serialize(self, buffer="std::vector<char>"):
        "Serialize to a buffer of bytes"
        return "void"

    @PYB11implementation("""
    [](uniform_random_01& self, std::vector<char>& buffer, size_t pos) {
      auto itr = buffer.begin() + pos;
      self.deserialize(itr, buffer.end());
      return std::distance(buffer.begin(), itr);
    }""")
    def deserialize(self,
                    buffer = "std::vector<char>",
                    pos = "size_t"):
        "Deserialize from a buffer of bytes, starting at the given position.  Returns the new position in the buffer."
        return "size_t"

    #...........................................................................
    # Operators
    def __eq__(self):
        return

    def __ne__(self):
        return

File: src/Utilities/uniform_random_01.cc
//---------------------------------Spheral++----------------------------------//
// uniform_random_01
//
// Encapsulate a random number generator to generate numbers in [0,1).
//
// We also require this generator be able to serialize/deserialize, both for
// restart and communication.
//
// Created by JMO, Mon May 10 16:02:11 PDT 2021
//----------------------------------------------------------------------------//

#include "uniform_random_01.hh"
#include "Utilities/packElement.hh"

namespace Spheral {

//------------------------------------------------------------------------------
// Constructors
//------------------------------------------------------------------------------
uniform_random_01::uniform_random_01():
  mGen(0u),
  mRan(0.0, 1.0),
  mSeed(0u),
  mNumCalls(0u) {
}

uniform_random_01::uniform_random_01(const size_t seed):
  mGen(seed),
  mRan(0.0, 1.0),
  mSeed(seed),
  mNumCalls(0u) {
}

uniform_random_01::uniform_random_01(const uniform_random_01& rhs):
  mGen(rhs.mGen),
  mRan(rhs.mRan),
  mSeed(rhs.mSeed),
  mNumCalls(rhs.mNumCalls) {
}

uniform_random_01&
uniform_random_01::operator=(const uniform_random_01& rhs) {
  if (this != &rhs) {
    mGen = rhs.mGen;
    mRan = rhs.mRan;
    mSeed = rhs.mSeed;
    mNumCalls = rhs.mNumCalls;
  }
  return *this;
}

//------------------------------------------------------------------------------
// Destructor
//------------------------------------------------------------------------------
uniform_random_01::~uniform_random_01() {
}


//------------------------------------------------------------------------------
// seed
//------------------------------------------------------------------------------
size_t
uniform_random_01::seed() const {
  return mSeed;
}

//------------------------------------------------------------------------------
// numCalls
//------------------------------------------------------------------------------
size_t
uniform_random_01::numCalls() const {
  return mNumCalls;
}

//------------------------------------------------------------------------------
// Set the seed
//------------------------------------------------------------------------------
void
uniform_random_01::seed(const size_t val) {
  mSeed = val;
  mGen.seed(val);
}

//------------------------------------------------------------------------------
// Advance the state number of steps, as though we generated that many random
// values.
//------------------------------------------------------------------------------
void
uniform_random_01::advance(const size_t n) {
  // mGen.discard(n * mGen.state_size);
  // mNumCalls += n;
  for (auto i = 0u; i < n; ++i) (*this)();
}

//------------------------------------------------------------------------------
// Serialize to a buffer
//------------------------------------------------------------------------------
void
uniform_random_01::serialize(std::vector<char>& buffer) const {
  packElement(mSeed, buffer);
  packElement(mNumCalls, buffer);
}

//------------------------------------------------------------------------------
// Deserialize from a buffer
//------------------------------------------------------------------------------
void
uniform_random_01::deserialize(std::vector<char>::const_iterator& itr,
                               const std::vector<char>::const_iterator endItr) {
  unpackElement(mSeed, itr, endItr);
  unpackElement(mNumCalls, itr, endItr);
  this->seed(mSeed);
  this->advance(mNumCalls);
}

}
